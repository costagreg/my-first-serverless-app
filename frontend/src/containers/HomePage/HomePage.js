import React, { useState } from 'react'
import UrlInputForm from '../../components/UrlInputForm'
import UrlShortenerOutput from '../../components/UrlShortenerOutput'
import { isValidUrl, sanitiseUrl } from '../../helpers/urlHelpers'
import axios from 'axios'


const HomePage = ({ appConfig = {} }) => {
  const { apiUrl } = appConfig
  const [error, setError] = useState('')
  const [urlShortened, setUrlShortened] = useState('')

  const createUrl = (url) => {

    if (isValidUrl(url)) {
      axios
        .post(`${apiUrl}/api/url`, {
          url: sanitiseUrl(url),
        })
        .then(({ data }) => {
          const { result } = data
          const { id } = result

          setUrlShortened(`${apiUrl}/${id}`)
        })
        .catch((error) => {
          // setError(error)
          console.log(error)
        })
    } else {
      setError('Url not valid')
    }
  }

  return (
    <div>
      <center>
        <h2>URL Shortener!</h2>
        <p>Simply copy & paste your link and click the cut button!</p>
      </center>
      <UrlInputForm createUrl={createUrl} error={error} />
      <UrlShortenerOutput error={error} urlShortened={urlShortened} />
    </div>
  )
}

export default HomePage
