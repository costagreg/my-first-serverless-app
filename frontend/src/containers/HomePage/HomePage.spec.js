import React from 'react'
import HomePage from './HomePage'
import { render, fireEvent, waitFor, screen } from '@testing-library/react'
import mockAxios from 'axios'

jest.mock('axios')

const props = {
  appConfig: { apiUrl: 'http://api.test.com' },
}

describe('HomePage', () => {
  describe('@renders', () => {
    it('renders the Dashboard without any error', () => {
      const { asFragment } = render(<HomePage {...props} />)

      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('@submit', () => {
    mockAxios.post.mockImplementation(() =>
      Promise.resolve({
        data: {
          result: { id: 'xysuk' },
        },
      })
    )

    it('should call the API to add the url', async () => {
      const { getByRole, getByText } = render(<HomePage {...props} />)

      const urlInput = getByRole('textbox')
      const submitButton = getByRole('button')

      fireEvent.change(urlInput, {
        target: { value: 'http://www.example.com/' },
      })

      fireEvent.click(submitButton)

      expect(mockAxios.post).toHaveBeenCalledWith(
        'http://api.test.com/api/url',
        {
          url: 'http://www.example.com/',
        }
      )

      await waitFor(() => getByText('http://api.test.com/xysuk'))
    })
  })
})
