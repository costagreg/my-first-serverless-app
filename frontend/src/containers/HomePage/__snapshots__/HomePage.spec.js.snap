// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage @renders renders the Dashboard without any error 1`] = `
<DocumentFragment>
  <div>
    <center>
      <h2>
        URL Shortener!
      </h2>
      <p>
        Simply copy & paste your link and click the cut button!
      </p>
    </center>
    <form
      class="urlInputForm"
    >
      <input
        class="urlInputForm__input"
        placeholder="https://www.example.com"
        type="text"
        value=""
      />
      <input
        class="urlInputForm__submit"
        type="submit"
        value="Cut me!"
      />
    </form>
    <div
      class="urlShortenerOutput"
    />
  </div>
</DocumentFragment>
`;
